File: content/snippets/typescript/interview_lcci/04/10_check_subtree.ts
class Node {
  constructor(
    public val: number,
    public left: Node | null = null,
    public right: Node | null = null
  ) {}
}

function same(a: Node | null, b: Node | null): boolean {
  if (!a && !b) return true;
  if (!a || !b) return false;
  return a.val === b.val && same(a.left, b.left) && same(a.right, b.right);
}

function isSubtree(t1: Node | null, t2: Node | null): boolean {
  if (!t2) return true;
  if (!t1) return false;
  if (same(t1, t2)) return true;
  return isSubtree(t1.left, t2) || isSubtree(t1.right, t2);
}

const t1 = new Node(3, new Node(4, new Node(1), new Node(2)), new Node(5));
const t2 = new Node(4, new Node(1), new Node(2));
const t3 = new Node(4, new Node(1), new Node(3));

console.log(isSubtree(t1, t2));
console.log(isSubtree(t1, t3));

File: content/snippets/scala2/interview_lcci/04/10_check_subtree.scala
object Lcci0410CheckSubtree {
  final case class Node(value: Int, var left: Node | Null = null, var right: Node | Null = null)

  def same(a: Node | Null, b: Node | Null): Boolean = {
    if (a == null && b == null) return true
    if (a == null || b == null) return false
    val x = a.asInstanceOf[Node]
    val y = b.asInstanceOf[Node]
    x.value == y.value && same(x.left, y.left) && same(x.right, y.right)
  }

  def isSubtree(t1: Node | Null, t2: Node | Null): Boolean = {
    if (t2 == null) return true
    if (t1 == null) return false
    val cur = t1.asInstanceOf[Node]
    same(cur, t2) || isSubtree(cur.left, t2) || isSubtree(cur.right, t2)
  }

  def main(args: Array[String]): Unit = {
    val t1 = Node(3, Node(4, Node(1), Node(2)), Node(5))
    val t2 = Node(4, Node(1), Node(2))
    val t3 = Node(4, Node(1), Node(3))
    println(isSubtree(t1, t2))
    println(isSubtree(t1, t3))
  }
}
